File: frontend/src/utils/mealSectionRules.ts
const BREAKFAST_CATEGORIES = ['早餐', '早点'] as const;
const LUNCH_CATEGORIES = ['饭盒', '大型供餐'] as const;
const DINNER_CATEGORIES = ['饭盒', '大型供餐'] as const;

export const getMealSectionCategoryRule = (name: string): readonly string[] | null => {
  const normalizedName = String(name || '').trim().toLowerCase();
  if (normalizedName.includes('早餐') || normalizedName.includes('breakfast')) {
    return BREAKFAST_CATEGORIES;
  }
  if (normalizedName.includes('午餐') || normalizedName.includes('lunch')) {
    return LUNCH_CATEGORIES;
  }
  if (normalizedName.includes('晚餐') || normalizedName.includes('dinner')) {
    return DINNER_CATEGORIES;
  }
  return null;
};

export const getDefaultMealSectionCategories = (name: string): string[] => {
  const normalizedName = String(name || '').trim().toLowerCase();
  if (normalizedName.includes('早餐') || normalizedName.includes('breakfast')) {
    return ['早餐'];
  }
  if (
    normalizedName.includes('午餐')
    || normalizedName.includes('lunch')
    || normalizedName.includes('晚餐')
    || normalizedName.includes('dinner')
  ) {
    return ['大型供餐'];
  }
  return [];
};

export const parseMealSectionCategories = (value: string | string[] | null | undefined): string[] => {
  const values = Array.isArray(value) ? value : String(value || '').split(',');
  return [...new Set(values.map(item => item.trim()).filter(Boolean))];
};

export const getEffectiveMealSectionCategories = (
  name: string,
  configuredCategories: string | string[] | null | undefined,
): string[] => {
  const configured = parseMealSectionCategories(configuredCategories);
  const rule = getMealSectionCategoryRule(name);
  return rule ? configured.filter(category => rule.some(allowed => allowed === category)) : configured;
};

export const getInvalidMealSectionCategories = (
  name: string,
  configuredCategories: string | string[] | null | undefined,
): string[] => {
  const configured = parseMealSectionCategories(configuredCategories);
  const rule = getMealSectionCategoryRule(name);
  return rule ? configured.filter(category => !rule.some(allowed => allowed === category)) : [];
};

File: backend/api/meal_section_rules.py
from collections.abc import Iterable

from model.models import MealSection


BREAKFAST_CATEGORIES = frozenset({"早餐", "早点"})
LUNCH_CATEGORIES = frozenset({"饭盒", "大型供餐"})
DINNER_CATEGORIES = frozenset({"饭盒", "大型供餐"})


def category_rule_for_meal_section(name: str) -> frozenset[str] | None:
    normalized_name = (name or "").strip().lower()
    if "早餐" in normalized_name or "breakfast" in normalized_name:
        return BREAKFAST_CATEGORIES
    if "午餐" in normalized_name or "lunch" in normalized_name:
        return LUNCH_CATEGORIES
    if "晚餐" in normalized_name or "dinner" in normalized_name:
        return DINNER_CATEGORIES
    return None


def parse_categories(value: str | Iterable[str] | None) -> set[str]:
    values = value.split(",") if isinstance(value, str) else (value or [])
    return {str(item).strip() for item in values if str(item).strip()}


def effective_meal_section_categories(section: MealSection) -> set[str]:
    configured = parse_categories(section.allowed_categories)
    rule = category_rule_for_meal_section(section.name)
    return configured.intersection(rule) if rule is not None else configured


def invalid_meal_section_categories(name: str, categories: str | Iterable[str] | None) -> set[str]:
    configured = parse_categories(categories)
    rule = category_rule_for_meal_section(name)
    return configured.difference(rule) if rule is not None else set()
